uberfire-extensions/uberfire-metadata/uberfire-metadata-backends/uberfire-metadata-backend-infinispan/src/main/java/org/uberfire/ext/metadata/backend/infinispan/utils/AttributesUtil.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright 2018 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.uberfire.ext.metadata.backend.infinispan.utils;
+
+public class AttributesUtil {
+
+    public static String toProtobufFormat(String value) {
+        return value.replaceAll("\\.",
```

Comment:
This implementation is rather inefficient for two reasons:
1. `String.replace()` is calling `Pattern.compile()` internally every time you call it which is several orders. Compiling patterns is several orders of magnitude slower than matching/replacing. When you know the pattern in advance, you should compile Pattern once and then create new Matcher every time you need to replace something in a string.

2. You calling it repetitively in a chain where you could actually just calling it once by specifying just one regex to replace. 

Based on the above I'd reimplement this like so:

```java
package org.uberfire.ext.metadata.backend.infinispan.utils;

import java.util.regex.Pattern;

public class AttributesUtil {

    // Characters that cannot be present in protobuf keys and will be replaced by double underscore
    private static final Pattern INVALID_CHAR = Pattern.compile("[./:\\[\\]]");
    private static final Pattern DOUBLE_UNDERSCORE = Pattern.compile("__");

    public static String toProtobufFormat(String value) {
        return INVALID_CHAR.matcher(value).replaceAll("__")
                //If '-' should be replaced with underscore, include it into the pattern ^
                .replaceAll("-", "_");
    }

    public static String toKPropertyFormat(String value) {
        return DOUBLE_UNDERSCORE.matcher(value).replaceAll("\\.");
    }
}

```

And here's a test as a bonus for you :)

```java
/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package org.uberfire.ext.metadata.backend.infinispan.utils;

import org.junit.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class AttributesUtilTest {

    @Test
    public void toProtobufFormat_replaceInvalidCharactersWithTwoUnderscores() {
        assertThat(AttributesUtil.toProtobufFormat("abcdEFGH")).isEqualTo("abcdEFGH");
        assertThat(AttributesUtil.toProtobufFormat("a.b")).isEqualTo("a__b");
        assertThat(AttributesUtil.toProtobufFormat("c/d")).isEqualTo("c__d");
        assertThat(AttributesUtil.toProtobufFormat("e:f")).isEqualTo("e__f");
        assertThat(AttributesUtil.toProtobufFormat("g[h]")).isEqualTo("g__h__");
        assertThat(AttributesUtil.toProtobufFormat("i-j")).isEqualTo("i_j");
        assertThat(AttributesUtil.toProtobufFormat("./:[]-")).isEqualTo("___________");
    }

    @Test
    public void toKPropertyFormat_replacesDoubleUnderscoreWithSlashDot() {
        assertThat(AttributesUtil.toKPropertyFormat("a__b_c__d")).isEqualTo("a.b_c.d");
    }
}

```